CMakeLists.txt

```diff
@@ -1,17 +1,83 @@
 cmake_minimum_required(VERSION 3.10)
 
-project(cxxeffect)
+project(cxxeffect CXX)
 
-# add the executable
-add_executable(cxxeffect
-    src/stream.hpp
-    src/stream.cpp
-    src/task.hpp
-    src/task.cpp
-    src/eff.hpp
-    src/eff.cpp)
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
+# cxxeffect is an interface library b/c it has a lot of templated code
+# That means it's best implemented as header-only without the need for cpp files
+add_library(cxxeffect INTERFACE)
 
-target_link_libraries(cxxeffect pthread)
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
+if (MSVC)
+    target_compile_options(cxxeffect INTERFACE "/std:c++latest" "/EHsc")
+endif()
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" )
+    target_Compile_options(cxxeffect INTERFACE "-std=c++20")
+endif()
+
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" )
+    target_compile_options(cxxeffect INTERFACE "-std=c++20" "-fcoroutines")
+endif()
+
+
+target_compile_definitions(cxxeffect INTERFACE)
+
+target_include_directories(
+    cxxeffect
+    INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/>
+    $<INSTALL_INTERFACE:include>
+)
+
+# Provides install target, so users can install the library on their system
+# if they like
+
+option(CXXEFFECT_Install "Install CMake targets during install step." ON)
+install(
+    DIRECTORY ${PROJECT_SOURCE_DIR}/include/
+    DESTINATION include
+)
+
+# Use pthreads, if availible
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+# Enable threads in project
+find_package(Threads REQUIRED)
+
+# If we're building this project, and not something that has this project as a dependency
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+    file(GLOB all_tests "examples/*.cpp")
+
+    foreach(file_name ${all_tests})
+        get_filename_component(executable_name ${file_name} NAME_WE)
+        add_executable("${executable_name}" "${file_name}")
+        target_link_libraries("${executable_name}" PRIVATE cxxeffect)
+        target_link_libraries("${executable_name}" PRIVATE Threads::Threads)
+    endforeach()
+endif()
```

Comment:
Is it possible to separate building the examples from building the header file (and eventually library)?

Reply:
Yes. This line here:
```cmake
if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
Ensures that examples only get built if you're building the project itself. If it's used as a dependency in someone else's project, examples won't get built. We could also add a flag so that examples are only built if the flag is enabled.

Anyone using the library can link it simply by doing
```cmake
        target_link_libraries("${executable_name}" PRIVATE cxxeffect)
```
And this works independent of cxxeffects being header-only or headers with shared object files, and cmake will do the correct thing in each case.

(That being said we have nothing to put in SO files right now)

Reply:
Alright. Imo it should be possible to skip the building of examples (some people don't use CMake and don't have a lot of time on their hands) to just build the SFH or the SO.